From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Leandro Dorileo <leandro.maciel.dorileo@intel.com>
Date: Thu, 19 Sep 2019 14:35:00 -0700
Subject: [PATCH] Add ldconfig post transaction hook.

Because autospec does not generate any %post. This also will reduce amount of calls
to ldconfig, when all the manual %post are dropped.

Ported-by: Leandro Dorileo <leandro.maciel.dorileo@intel.com>
Signed-off-by: Thiago Macieira <thiago.macieira@intel.com>
---
 macros.in           |  1 +
 plugins/Makefile.am |  4 ++++
 plugins/ldconfig.c  | 36 ++++++++++++++++++++++++++++++++++++
 3 files changed, 41 insertions(+)
 create mode 100644 plugins/ldconfig.c

diff --git a/macros.in b/macros.in
index a6069ee..bf0fb2e 100644
--- a/macros.in
+++ b/macros.in
@@ -1174,6 +1174,7 @@ package or when debugging this package.\
 %__transaction_syslog		%{__plugindir}/syslog.so
 %__transaction_ima		%{__plugindir}/ima.so
 %__transaction_prioreset	%{__plugindir}/prioreset.so
+%__transaction_ldconfig		%{__plugindir}/ldconfig.so
 
 #------------------------------------------------------------------------------
 # Macros for further automated spec %setup and patch application
diff --git a/plugins/Makefile.am b/plugins/Makefile.am
index ab4eee3..120f456 100644
--- a/plugins/Makefile.am
+++ b/plugins/Makefile.am
@@ -37,6 +37,10 @@ syslog_la_SOURCES = syslog.c
 syslog_la_LIBADD = $(top_builddir)/lib/librpm.la $(top_builddir)/rpmio/librpmio.la
 plugins_LTLIBRARIES += syslog.la
 
+ldconfig_la_SOURCES = ldconfig.c
+ldconfig_la_LIBADD = $(top_builddir)/lib/librpm.la $(top_builddir)/rpmio/librpmio.la
+plugins_LTLIBRARIES += ldconfig.la
+
 if IMA
 ima_la_sources = ima.c
 ima_la_LIBADD = $(top_builddir)/lib/librpm.la $(top_builddir)/rpmio/librpmio.la
diff --git a/plugins/ldconfig.c b/plugins/ldconfig.c
new file mode 100644
index 0000000..5ab54ba
--- /dev/null
+++ b/plugins/ldconfig.c
@@ -0,0 +1,36 @@
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <rpm/rpmlog.h>
+#include <rpm/rpmts.h>
+#include "lib/rpmplugin.h"
+
+static rpmRC ldconfig_tsm_post(rpmPlugin plugin, rpmts ts, int res)
+{
+    int pid = -1;
+    siginfo_t info;
+    do {
+	pid = fork ();
+    } while (pid < 0);
+
+    if (pid) {
+	info.si_code = 0;
+	info.si_status = 0;
+
+	if (waitid (P_PID, pid, &info, WEXITED)) {
+	    rpmlog(RPMLOG_DEBUG, "Failed to wait for ldconfig\n");
+	}
+
+	if (info.si_code != CLD_EXITED || info.si_status) {
+	    rpmlog(RPMLOG_DEBUG, "Bad ldconfig exit code %i and status %i\n", info.si_code, info.si_status);
+	}
+
+    } else {
+	execlp ("ldconfig", "ldconfig", (char *)NULL);
+    }
+
+    return RPMRC_OK;
+}
+
+struct rpmPluginHooks_s ldconfig_hooks = {
+    .tsm_post = ldconfig_tsm_post,
+};
